### DISCUSSION TITLE
Elastic response to mesh modification

### DISCUSSION THREAD

**Discussion by wehrad** [idaholab/moose#20836] [Q&A Getting Started] [answered]
Hi everyone,

I have been working on an application based on MASTODON to study the elastic response of a domain on an inclined surface, after the removal of some support at its front.

What I did first was to apply a Pressure boundary condition, with the `frontal_pressure` function depending on time and taking two different values before and after a given time t (simulating the loss of support in front of the domain). 

```
[BCs]
  [Pressure]
    [front_pressure]  
    boundary = front
    function = frontal_pressure
    displacements = 'disp_x disp_y disp_z'
    []
  []
```

This worked great and gave interesting results. But this only represents the result of a material removal, not the actual removal itself. Next step was therefore to actively remove some material at the front of the domain and get the pressure release as a result, instead of a preset. Just as it actually happens in a real case. 

Based on https://github.com/idaholab/moose/discussions/20601, I looked into two potential solutions:
1) https://mooseframework.inl.gov/source/userobject/CoupledVarThresholdElementSubdomainModifier.html
2) https://mooseframework.inl.gov/source/meshgenerators/BlockDeletionGenerator.html

My progress (and issues) on those two options:

1) This is on the fly. One input file, and a moving boundary. However, I need to assign the newly deactivated block (considered as removed, chopped off) to an existing subdomain. 
And that is where I got stuck... How to assign it to a subdomain that doesn't have **any**, literally any, impact on the main domain anymore? 
Here is how far I got into that first direction with my input file: [chop_dynamic.i](https://github.com/AdrienWehrle/moose/blob/next/examples/ex23_chop/chop_dynamic.i)

2) This is the restart option. Two input files, the first one for initialization, and the second one where a new mesh is created with the material chopped off and the solutions remapped based on the first simulation. It also seems like a good idea, because in my case I remove material only once and this is irreversible. 
I think I got pretty far here, but somehow I don't get significant elastic response (only a very slow displacement in x) in my transient simulation although I removed 500m off my domain and initialized with displacements corresponding to the entire loaded domain... So I should have a significant extension-dominated reaction...! Also the solve is very slow, much slower than with a pressure release of similar magnitude, somehow suggesting something is wrong...
Here is how far I got into that second direction with my input files: [chop_init.i](https://github.com/AdrienWehrle/moose/blob/next/examples/ex23_chop/chop_init.i) (initialization) and [chop_restart.i](https://github.com/AdrienWehrle/moose/blob/next/examples/ex23_chop/chop_restart.i) (transient after removal) 

Any help/thoughts on one of these options or both would be much appreciated! Thank you a lot in advance!

**Comment by GiudGiud** [comment, marked as answer]
Hello

For option 1, if you move elements to a domain that is outside the block restrictions of the kernels/variables, then it is effectively removed from the simulation. 
There might also be material property values you could set in that domain that negate all form of displacement (like incredibly stiff). This might not be a good idea as I think you may pay a steep price in convergence properties. 

For your inputs in 1 and 2, I would set displacement = 'disp_x disp_y ...' in GlobalParams. it s too easy to have forgotten it somewhere. 

Guillaume

**Reply by wehrad** [reply]
Hi @GiudGiud ! Thank you a lot for your help!

> For option 1, if you move elements to a domain that is outside the block restrictions of the kernels/variables, then it is effectively removed from the simulation.

It sounds like option 1 is the better option here, so I tried to move elements to a domain outside of the block restrictions of my problem. To this end, I think I need to create some elements disconnected from my domain, defined in [channel.e](https://github.com/AdrienWehrle/moose/blob/next/examples/ex23_chop/channel.e).

I therefore tried the following in [chop_dynamic.i](https://github.com/AdrienWehrle/moose/blob/next/examples/ex23_chop/chop_dynamic.i):

```
[Mesh]
  [channel]      
  type = FileMeshGenerator
  file = channel.e
  []
  
  [chop_event]
    type = SubdomainBoundingBoxGenerator
    input = channel
    block_id = 2
    block_name = chop_event
    bottom_left = '29500  -100 0'
    top_right = '30000 2100 1000'
  []

  [deactivated]
    type = SubdomainBoundingBoxGenerator
    input = calving_event
    block_id = 3
    block_name = deactivated_domain
    bottom_left = '40000  1000 1000'
    top_right = '40001  1001 1001'
  []
[]
```
with the `CoupledVarThresholdElementSubdomainModifier` `subdomain_id` set to 3. And where `deactivated` has somehow random coordinates, just to be out of the main domain boundaries.

However I get `Requested subdomain 3 does not exist.`. I suppose the subdomain wasn't created because no element are in the bounding box. I looked at other objects in https://mooseframework.inl.gov/syntax/Mesh/index.html but I couldn't find the one that I should use here so far... Or maybe am I not interpreting your advice the right way? Thank you a lot for your help!

**Reply by GiudGiud** [reply]
hello

You could use an ElementGenerator and CombinerGenerator for this:
https://mooseframework.inl.gov/source/meshgenerators/ElementGenerator.html
https://mooseframework.inl.gov/source/meshgenerators/CombinerGenerator.html

just put an element real far and block restrict the physics out of it. 

Guillaume

**Reply by wehrad** [reply]
Thank you a lot! What don't get right with https://mooseframework.inl.gov/source/meshgenerators/ElementGenerator.html is that it only generates an element if I use existing nodes for `element_connectivity`, although I want to create an element that is not connected at all to my main domain, hence with new nodes... 

Here is the mesh I'm working on:

```

[Mesh]
  
  [channel]      
  type = FileMeshGenerator
  file = channel.e
  []

  [chop_event]
    type = SubdomainBoundingBoxGenerator
    input = channel
    block_id = 2
    block_name = chop_event
    bottom_left = '29500  -100 0'
    top_right = '30000 2100 1000'
  []

  [single_element]
    type = ElementGenerator
    input = chop_event
    nodal_positions = '0 0 0
                       -1000 0 0
                       -1000 -1000 0
                       0 -1000 0
                       0 0 -500
                       -1000 0 -500
                       -1000 -1000 -500
                       0 -1000 -500'
    element_connectivity = '100000 100001 100002 100003 100005 100006 100007 100008'
    elem_type = "QUAD8"
  []

  [combined]
    type = CombinerGenerator
    inputs = 'chop_event single_element'
  []

  final_generator = combined
[]
```

I set `element_connectivity` higher than my maximum node number to try to create new ones, but it fails with a segmentation fault (because those nodes don't exist).... If I set `element_connectivity` in my range of existing nodes, then it uses those... Somehow I'm using it in the wrong way since it doesn't allow me make it do what's meant to...! 

Thank you a lot for your help!

**Reply by GiudGiud** [reply]
umm. Well if it's not in a block that has the physics it shouldnt matter at all where it is. Or how many elements it has

If that mesh generator doesnt work for you you can use any random mesh, load it with FileMeshGenerator and combine it, unstitched, with the CombinerMeshGenerator. And it ll just sit there, doing nothing. 

You'll need this btw
```
[Problem]
  kernel_coverage_check = false
[]
```

**Reply by wehrad** [reply]
Thank you a lot for you help on that @GiudGiud! I think I'm getting close, but still have an issue.
In the updated version of [chop_dynamic.i](https://github.com/AdrienWehrle/moose/blob/next/examples/ex24_chop/chop_dynamic.i), I:
- Load my main domain + a single isolated element
- Set a new block id for the isolated element only (value is 255)
- Combine the two meshes
- Restrict all the physics to blocks 1 and 2 (main mesh and area that I want to chop off, respectively)
- Change block 2 to 255 at the chopping time to move the chopped area out of my problem.
- Set `kerner_coverage_check` and `material_coverage_check` to False.

It somehow sounds good but I get the following error, which is linked to the element doing nothing at the solve level but I can't really backtrack it nor solve it (error is at the end of the log):
```
The following total 1 aux variables:
  von_mises
are added for automatic output by MaterialOutputAction.
Finished Setting Up                                                                      [  0.65 s] [  187 MB]
Framework Information:
MOOSE Version:           git commit 6f309b1abd on 2022-02-07
LibMesh Version:         
PETSc Version:           3.15.1
SLEPc Version:           3.15.1
Current Time:            Tue Apr 26 09:30:20 2022
Executable Timestamp:    Fri Mar 11 16:19:52 2022

Parallelism:
  Num Processors:          1
  Num Threads:             11

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   11987
  Elems:                   9864
  Num Subdomains:          3

Nonlinear System:
  Num DOFs:                35961
  Num Local DOFs:          35961
  Variables:               { "disp_x" "disp_y" "disp_z" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                182549
  Num Local DOFs:          182549
  Variables:               { "vel_x" "accel_x" "vel_y" "accel_y" "vel_z" "accel_z" } { "stress_xx" "stress_xy" 
                             "stress_xz" "stress_yx" "stress_yy" "stress_yz" "stress_zx" "stress_zy" 
                             "stress_zz" } "t_calving" "von_mises" 
  Finite Element Types:    "LAGRANGE" "MONOMIAL" "LAGRANGE" "MONOMIAL" 
  Approximation Orders:    "FIRST" "CONSTANT" "FIRST" "CONSTANT" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP

LEGACY MODES ENABLED:
 This application uses the legacy material output option: material properties are output only on TIMESTEP_END, not INITIAL. To remove this message, set 'use_legacy_material_output' to false in this application. If there are gold output files that contain material property output for which output occurs on INITIAL, then these will generate diffs due to zero values being stored, and these tests should be re-golded.


    Setting Up Materials
      Finished Computing Initial Material Values                                         [  0.21 s] [  131 MB]
    Finished Setting Up Materials                                                        [  0.21 s] [  131 MB]
  Finished Performing Initial Setup                                                      [  2.77 s] [  185 MB]

Time Step 0, time = 0
Warning, Exodus files cannot have titles longer than 80 characters.  Your title will be truncated.

Time Step 1, time = 0.02, dt = 0.02

Performing automatic scaling calculation

 0 Nonlinear |R| = 9.256317e-01
At column 2, pivotL() encounters zero diagonal at line 722 in file /opt/civet/build_2/conda_builds/conda_envs/next-aeb8c4f62f/conda-bld/moose-petsc_1643733610091/work/arch-conda-c-opt/externalpackages/git.superlu_dist/SRC/symbfact.c
```

Thank you a lot for your precious help!

**Reply by GiudGiud** [reply]
oh did you block restrict the variables outside of the elements with no physics? This is actually important, otherwise the matrix has a zero for the qps there. 

The block restriction of the variables is usually inherited by the kernels btw

**Reply by wehrad** [reply]
> oh did you block restrict the variables outside of the elements with no physics?

I am not sure I completely understand. Variables are not block-restricted at the moment, but they probably should? Should I use `inactive` on block 255 (inactive element) for all variables?

**Reply by GiudGiud** [reply]
They should be. Pass the `block = ` in the variables blocks (each of them) not the kernels. 

inactive is not for this. inactive is to turn off an object in the input file, like entirely not just block restrict

**Reply by wehrad** [reply]
Thank you a lot, that indeed solved the zero qps issue! 

Only remaining problem is I restrict the physics to block 1 and 2 (main domain and chopped area, respectively), then change the block ID of the chopped area from 2 to 255 (to deactivate it). Therefore block 2 doesn't exist anymore, but is passed in `block =` in variable blocks, resulting in `ExodusII_IO_Helper: block id 2 not found in block_ids.` at chop time (t=1). 
I think one workaround is to have not only the one isolated element (block_id=255), but also a second one with block_id=2 (useless but containing the physics) so that block id 2 doesn't disappear. Do you think that is a good idea?

Thank you a lot for your help!

**Reply by GiudGiud** [reply]
Yeah I think you ll have to do that. 
Or you start with just all elements in block 1, then move some of the chopped elements in block 1 to block 255?

**Comment by wehrad** [comment]
> Or you start with just all elements in block 1, then move some of the chopped elements in block 1 to block 255?

Indeed, I managed to trade my initial `SubdomainBoundingBoxGenerator` + `CoupledVarThresholdElementSubdomainModifier` with threshold on time, with one single `CoupledVarThresholdElementSubdomainModifier` on a variable depending on space and time! Thank you so much for your help on that!

**Comment by wehrad** [comment]
One last thing that I can't get right for this problem (sorry to disturb you with that!):
I added the following function:
```
[Functions]
  [chop_criterion]
    type = ParsedFunction
    value = 'if((x>29300.)&(t>1.), 1., 0.)'
  []
[]
```

that I then apply in my subdomain modifier:

```
[UserObjects]
  [calving_event]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = 'chop_boolean'
    block = 1
    criterion_type = ABOVE
    threshold = 0.5
    subdomain_id = 255
    moving_boundary_name = downstream 
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]
```
`chop_boolean` is the variable set by `chop_criterion`.

I thought I solved my issue, but I get a weird chop where not all expected elements are moved to the 255 subdomain (see screenshots below) although the condition is a clear cut and not at the border of an element row (x=29000 and x=29500 are the two element rows before and after the condition x=29300)... As you can see on the sfcreenshots, the outer band of elements is entirely removed but not issue appears in the second one....

This is independent of the threshold, I always get such elements in and out no matter what value I take...

I was earlier using the same spatial condition with `SubdomainBoundingBoxGenerator` and I got a nice and clear cut as expected... Any ideas on that? Thank you a lot

![chop4](https://user-images.githubusercontent.com/44843368/165945248-079910d3-4682-4a73-b791-abba9deecb93.png)
![chop5](https://user-images.githubusercontent.com/44843368/165945253-576f0d49-eb42-475a-89f3-52d6738ad341.png)

**Reply by GiudGiud** [reply]
I think the centroid of the element is considered. If you move this x=29300 closer to the next layer (3rd) does it cut properly?

**Reply by GiudGiud** [reply]
actually no scratch that.
This is how the element value is determined:
```
  Real avg_val = 0;

  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
    avg_val += _v[qp] * _JxW[qp] * _coord[qp];
  avg_val /= _current_elem_volume;

  return avg_val;
```

sp you have to make that average below the threshold value.
Either smooth the transition with a linear function in between the jump, or jump to -1 instead of 0, or raise/lower the threshold appropriately

**Reply by wehrad** [reply]
Oh woh, I wasn't looking in the right direction, I also initially thought it was about the spatial threshold. 

What I have for the moment is my chop_criterion taking -1000 or 1000 and threshold at 0 so there can't be any overrunning.

Thank you a lot for all your help on that! :+1:

**Reply by wehrad** [reply]
Next step will be to try and have my `chop_criterion` to depend on a damage value I compute from stresses. Challenge is my damage is a [material property](https://github.com/AdrienWehrle/moose/blob/next/examples/ex22_customdamage/custom_damage_application.i#L280) and not an (aux)variable, therefore I can't access it directly with a `CoupledVarThresholdElementSubdomainModifier`... I tried to find a way to convert a material property to a variable, but that sounds like a bad idea. Would you  by any chance have an advice on that before I mark this question as answered? Thank you a lot!

The damage implementation is visible [here](https://github.com/AdrienWehrle/moose/blob/next/examples/ex22_customdamage/custom_damage_application.i).

**Reply by GiudGiud** [reply]
Hello

it s fine to convert material property to auxiliary variable in your case. You are only interested about the value. 
it s not fine when you care about transferring derivatives. 

To do that use MaterialRealAux or one of the other template instantiations depending on the type of your material property 

Guillaume
